src/rgb/contract/conceal.rs

```diff
@@ -22,21 +23,77 @@ pub trait AutoConceal {
     fn conceal_except(&mut self, seals: &Vec<seal::Confidential>) -> usize;
 }
 
-// TODO: Do an auto implementation for and other collection types
-
 impl<T> AutoConceal for Vec<T>
 where
     T: AutoConceal + Ord + Clone,
 {
     fn conceal_except(&mut self, seals: &Vec<seal::Confidential>) -> usize {
         let mut count = 0;
-        let mut new_self = Vec::<T>::new();
+        for item in self.iter_mut() {
+            count += item.conceal_except(seals);
+        }
+        count
+    }
+}
+
+impl<T> AutoConceal for BTreeSet<T>
+where
+    T: AutoConceal + Ord + Clone,
+{
+    fn conceal_except(&mut self, seals: &Vec<seal::Confidential>) -> usize {
+        let mut count = 0;
+        let mut new_self = BTreeSet::<T>::new();
+        for item in self.iter() {
+            let mut new_item = item.clone();
+            count += new_item.conceal_except(seals);
+            new_self.insert(new_item);
+        }
+        *self = new_self;
+        count
+    }
+}
+
+impl<K, V> AutoConceal for BTreeMap<K, V>
+where
+    V: AutoConceal + Ord + Clone,
+    K: Ord + Clone,
+{
+    fn conceal_except(&mut self, seals: &Vec<seal::Confidential>) -> usize {
+        let mut count = 0;
+        for (_, v) in self.iter_mut() {
+            count += v.conceal_except(seals);
+        }
+        count
+    }
+}
+
+impl<T> AutoConceal for HashSet<T>
+where
+    T: AutoConceal + Ord + Clone + std::hash::Hash,
+{
+    fn conceal_except(&mut self, seals: &Vec<seal::Confidential>) -> usize {
+        let mut count = 0;
+        let mut new_self = HashSet::<T>::new();
         for item in self.iter() {
             let mut new_item = item.clone();
             count += new_item.conceal_except(seals);
-            new_self.push(new_item);
+            new_self.insert(new_item);
         }
         *self = new_self;
         count
     }
 }
+
+impl<K, V> AutoConceal for HashMap<K, V>
+where
+    V: AutoConceal + Ord + Clone + std::hash::Hash,
```

Comment:
Not sure why we need value here to implement `Hash` (it's only required for keys). Also `Ord` is unnecessary for both key and value in `HashMap` and values of `HashSet` (this is needed only for keys in `BTreeMap` and values of `BTreeSet`)

Reply:
`Ord` and `Hash` traits are required because without them `insert()` doesn't work.